presto-main/src/test/java/com/facebook/presto/sql/planner/PlanAssert.java

```diff
@@ -0,0 +1,415 @@
+    {
+        return anyNode(sources).with((node, session, metadata, symbolAliases) -> {
+            if (node instanceof JoinNode) {
+                JoinNode joinNode = (JoinNode) node;
+                if (joinNode.getCriteria().size() == expectedEquiCriteria.size()) {
+                    int i = 0;
+                    for (EquiJoinClause equiJoinClause : joinNode.getCriteria()) {
+                        AliasPair expectedEquiClause = expectedEquiCriteria.get(i++);
+                        symbolAliases.put(expectedEquiClause.left, equiJoinClause.getLeft());
+                        symbolAliases.put(expectedEquiClause.right, equiJoinClause.getRight());
+                    }
+                    return true;
+                }
+            }
+            return false;
+        });
+    }
+
+    public static AliasPair aliasPair(String left, String right)
+    {
+        return new AliasPair(left, right);
+    }
+
+    public static PlanMatcher filterNode(String predicate, PlanMatcher... sources)
+    {
+        Expression expectedPredicate = new SqlParser().createExpression(predicate);
+        return anyNode(sources).with((node, session, metadata, symbolAliases) -> {
+            if (node instanceof FilterNode) {
+                FilterNode filterNode = (FilterNode) node;
+                if (new ExpressionVerifier(symbolAliases).process(filterNode.getPredicate(), expectedPredicate)) {
+                    return true;
+                }
+            }
+            return false;
+        });
+    }
+
+    private static class ExpressionVerifier
+            extends AstVisitor<Boolean, Expression>
+    {
+        private final SymbolAliases symbolAliases;
+
+        public ExpressionVerifier(SymbolAliases symbolAliases)
+        {
+            this.symbolAliases = requireNonNull(symbolAliases, "symbolAliases is null");
+        }
+
+        @Override
+        protected Boolean visitNode(Node node, Expression context)
+        {
+            throw new IllegalStateException(format("Node %s is not supported", node));
+        }
+
+        @Override
+        protected Boolean visitNotExpression(NotExpression actual, Expression context)
+        {
+            if (context instanceof NotExpression) {
+                NotExpression expected = (NotExpression) context;
+                return process(actual.getValue(), expected.getValue());
+            }
+            return false;
+        }
+
+        @Override
+        protected Boolean visitQualifiedNameReference(QualifiedNameReference actual, Expression context)
+        {
+            if (context instanceof QualifiedNameReference) {
+                QualifiedNameReference expected = (QualifiedNameReference) context;
+                symbolAliases.put(expected.getName().toString(), Symbol.fromQualifiedName(actual.getName()));
+                return true;
+            }
+            return false;
+        }
+    }
+
+    public static final class PlanMatcher
```

Comment:
I think this code would be easier to read if these inner classes were moved to the top-level (at least, the public ones). I'd create a package "com.facebook.presto.sql.planner.assert" and move everything there.